help: document DetectSilence w/ multiple inputs

snappizz opened the pull request:
Please check my work here. This might be totally wrong, or there might be a better way to do this.

brianlheim commented on HelpSource/Classes/DetectSilence.schelp:
```diff
@@ -77,3 +78,15 @@ Task({
+
+section::DetectSilence and multiple inputs
+
+DetectSilence has a non-obvious interaction with multiple inputs. This code
```
I'd avoid saying 'non-obvious' here; this behavior could be perfectly predictably to someone who hasn't used DetectSilence before and is only familiar with multi-channel expansion. "Note that this line" should be enough.

brianlheim commented:
I think you want `FreeSelf` instead of `FreeSelfWhenDone`. `DetectSilence` doesn't have a 'done' flag detectable by `Done`/`FreeSelfWhenDone` (or if it does, that's not documented, which is another issue). At the very least, when I run your suggested workaround, it doesn't ever free itself.

My question is, couldn't one also use other one-liners like `sig.abs.sum`? That's been my workaround in the past. Obviously it's not a perfect equivalent to the workaround you suggest—in the sense that the threshold value now has a different meaning—but I don't think anyone's being that picky about their uses of `DetectSilence`. Here's a comparison of multiple approaches along with the number of UGens they spawn for an array of size 20.

```supercollider
(
// 1. 49 ugens
// naïve solution, will exit early
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig, doneAction:0);
	sig;
}.play

// 2. 37 ugens
// better, but will exit early if there's near-perfect phase cancellation
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.sum, doneAction:0);
	sig;
}.play

// 3. 57 ugens
// will not exit even with phase cancellation
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.abs.sum, doneAction:0);
	sig;
}.play

// 4. 58 ugens
// scaled by the number of channels so that the threshold value makes a little more sense
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	DetectSilence.ar(sig.abs.sum/sig.size, doneAction:0);
	sig;
}.play

// 5. 69 ugens
// most logically sound workaround
{
	var sig = ([SinOsc.ar, Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelf.kr(done.product);
	sig;
}.play

// 6. 70 ugens
// just to confirm that FreeSelf will actually end the synth
{
	var sig = ([Decay.ar(Impulse.ar(0), 2), Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelf.kr(done.product);
	sig;
}.play

// 7. 70 ugens
// doesn't free self
{
	var sig = ([Decay.ar(Impulse.ar(0), 2), Decay.ar(Impulse.ar(0), 1)]!10).flat;
	var done = DetectSilence.ar(sig);
	FreeSelfWhenDone.kr(done.product);
	sig;
}.play
)
```

Maybe we could give a few possible solutions and briefly discuss the caveats of each. Either way, thanks--this is a really good tip to document!

snappizz commented:
Good catch, @brianlheim. I've corrected it and added a note with the full-wave rectification solution.

I don't think we should go into much detail comparing solutions. If the user cares that much about the differences, they're probably knowledgeable enough to make the choice on their own.

brianlheim commented:
@snappizz Good point, works for me!

brianlheim commented on HelpSource/Classes/DetectSilence.schelp:
```diff
@@ -77,3 +78,17 @@ Task({
+
+code::FreeSelf.kr(DetectSilence.ar([left, right] + Impulse.ar(0)).product)::
+
+It's also possible to mix the signals with full-wave rectification to avert phase cancellation issues (code::[left, right].abs.sum / 2.sqrt::).
```
Just one tiny change - missing newline at the end of file for git's sake.

telephon commented:
btw. `sig.abs.sum/sig.size` is easier written as `sig.abs.mean`.

brianlheim commented:
Hah, I completely forgot that while I was testing this.

brianlheim commented:
@snappizz That makes sense but I'm having trouble verifying for myself. Replacing `sig.abs.sum` with `Mix.ar(sig.abs)` makes no difference in the server load or number of UGens spawned for me, even when I increase the size of the signal array to 200. Am I missing something?

telephon commented:
Both `sum` and `Mix` perform optimizations behind the scenes. I assumed that the only difference is that `a.sum`, where `a` is not an array, will fail.

(It would be reasonable to implement `aUGen.sum` and `aSimpleNumber.sum` each to return itself).

snappizz commented:
aha. So Mix is actually unnecessary?

telephon commented:
Here is a test:

```

x = { |n| SynthDef(\test, { Out.ar(0, Mix({ SinOsc.ar }.dup(n))) }) };
y = { |n| SynthDef(\test, {  Out.ar(0, { SinOsc.ar }.dup(n).sum) }) };

x.(1).children
y.(1).children

x.(2).children
y.(2).children

x.(3).children
y.(3).children

x.(4).children
y.(4).children

x.(5).children
y.(5).children

```

All are the same.

One difference is:

```
x = { |n| SynthDef(\test, { Mix({ SinOsc.ar }.dup(n)) }) };
y = { |n| SynthDef(\test, { { SinOsc.ar }.dup(n).sum }) };

x.(3).children // [ a SinOsc, a SinOsc, a SinOsc, a Sum3 ]
y.(3).children // nil
```
So when using `sum`, the pure ugens (without side effects) are getting optimized away.

For non-pure ugens, the optimization is correct:

```
y = { |n| SynthDef(\test, { { Line.kr }.dup(n).sum }) };
y.(3).children // [ a Line, a Line, a Line ]
```

telephon commented:
So I would propose that `Mix:new`should be implemented as:

```
*new { |array|
    ^array.asArray.sum
}
```

telephon commented:
… although the result can sometimes be better when you use `sum`:

```

x = { |n| SynthDef(\test, { Out.kr(0, Mix(f.dup(n))) }) };
y = { |n| SynthDef(\test, {  Out.kr(0, f.dup(n).sum) }) };


f = { |i| [2, 3, 4, 5, 6, Line.kr, SinOsc.ar] @@ i };
x.(6).children.collect { |x| [x, x.rate] }.do(_.postln)
y.(6).children.collect { |x| [x, x.rate] }.do(_.postln)


f = { |i| [ SinOsc.ar, 2, 3, 4, 5, 6, Line.kr, SinOsc.ar] @@ i };
x.(6).children.collect { |x| [x, x.rate] }.do(_.postln)
y.(6).children.collect { |x| [x, x.rate] }.do(_.postln)

``

Here, `sum` just sums numbers instead of making a `Sum3` UGe for them.

snappizz commented:
Discussion on Mix is cont'd at #2642.

brianlheim commented on HelpSource/Classes/DetectSilence.schelp:
```diff
@@ -77,3 +78,17 @@ Task({
+
+Be careful feeding multiple inputs into DetectSilence, since multichannel expansion can lead to confusing behavior. This code
+
+code::DetectSilence.ar([left, right] + Impulse.ar(0), doneAction: 2)::
```
Did you intend this to appear as runnable code? That only happens if `code::` and the closing `::` are on individual lines:
```
code::
DetectSilence...
::
```

snappizz replied:
It is not runnable code since it isn't a complete example.

brianlheim replied:
D'oh

brianlheim commented:
Ping. @snappizz Do you want to make that change re: `code::`? Are we ready to merge otherwise (I vote yes)?

snappizz commented:
Ready to merge.
